In [1]:
import pandas as pd, numpy as np
import statsmodels
from bokeh.plotting import figure, output_notebook, show
import sqlite3 as sql
output_notebook()

Loading BokehJS ...

In [2]:
con = sql.connect('E:/dev/CrpytoManager/CryptoDB.db')

Test with Bitcoin first

In [3]:
btc=pd.read_sql('SELECT * FROM btc_minute',con,index_col='datetime_utc')

In [4]:
btc.head()

open       high        low      close    volume  \
datetime_utc                                                                
2017-09-25 20:21:00  4316000.0  4321000.0  4316000.0  4321000.0  0.001888   
2017-09-25 20:22:00  4319000.0  4319000.0  4319000.0  4319000.0  0.000805   
2017-09-25 20:23:00  4318000.0  4318000.0  4318000.0  4318000.0  0.004647   
2017-09-25 20:24:00  4318000.0  4318000.0  4318000.0  4318000.0  0.000346   
2017-09-25 20:26:00  4318000.0  4318000.0  4318000.0  4318000.0  0.001326   

                     volume_amount  
datetime_utc                        
2017-09-25 20:21:00       8151.393  
2017-09-25 20:22:00       3476.795  
2017-09-25 20:23:00      20065.746  
2017-09-25 20:24:00       1494.028  
2017-09-25 20:26:00       5725.668

In [5]:
btc.index = pd.to_datetime(btc.index)

In [6]:
btc['price'] = btc[['open', 'high', 'low', 'close']].mean(axis=1)
btc.head()

open       high        low      close    volume  \
datetime_utc                                                                
2017-09-25 20:21:00  4316000.0  4321000.0  4316000.0  4321000.0  0.001888   
2017-09-25 20:22:00  4319000.0  4319000.0  4319000.0  4319000.0  0.000805   
2017-09-25 20:23:00  4318000.0  4318000.0  4318000.0  4318000.0  0.004647   
2017-09-25 20:24:00  4318000.0  4318000.0  4318000.0  4318000.0  0.000346   
2017-09-25 20:26:00  4318000.0  4318000.0  4318000.0  4318000.0  0.001326   

                     volume_amount      price  
datetime_utc                                   
2017-09-25 20:21:00       8151.393  4318500.0  
2017-09-25 20:22:00       3476.795  4319000.0  
2017-09-25 20:23:00      20065.746  4318000.0  
2017-09-25 20:24:00       1494.028  4318000.0  
2017-09-25 20:26:00       5725.668  4318000.0

In [7]:
btc['log_diff']=np.log(btc['price']).diff()
btc.head()

open       high        low      close    volume  \
datetime_utc                                                                
2017-09-25 20:21:00  4316000.0  4321000.0  4316000.0  4321000.0  0.001888   
2017-09-25 20:22:00  4319000.0  4319000.0  4319000.0  4319000.0  0.000805   
2017-09-25 20:23:00  4318000.0  4318000.0  4318000.0  4318000.0  0.004647   
2017-09-25 20:24:00  4318000.0  4318000.0  4318000.0  4318000.0  0.000346   
2017-09-25 20:26:00  4318000.0  4318000.0  4318000.0  4318000.0  0.001326   

                     volume_amount      price  log_diff  
datetime_utc                                             
2017-09-25 20:21:00       8151.393  4318500.0       NaN  
2017-09-25 20:22:00       3476.795  4319000.0  0.000116  
2017-09-25 20:23:00      20065.746  4318000.0 -0.000232  
2017-09-25 20:24:00       1494.028  4318000.0  0.000000  
2017-09-25 20:26:00       5725.668  4318000.0  0.000000

In [14]:
btc = btc.loc['2017-09-27': ' 2018-04-26']

In [15]:
btc.head()

open       high        low      close    volume  \
datetime_utc                                                                
2017-09-27 00:00:00  4322000.0  4322000.0  4322000.0  4322000.0  0.002642   
2017-09-27 00:01:00  4328000.0  4328000.0  4328000.0  4328000.0  0.003083   
2017-09-27 00:02:00  4323000.0  4323000.0  4323000.0  4323000.0  0.002071   
2017-09-27 00:03:00  4323000.0  4323000.0  4323000.0  4323000.0  0.000485   
2017-09-27 00:04:00  4320000.0  4323000.0  4318000.0  4323000.0  0.007036   

                     volume_amount      price  log_diff  
datetime_utc                                             
2017-09-27 00:00:00      11418.724  4322000.0  0.000231  
2017-09-27 00:01:00      13343.224  4328000.0  0.001387  
2017-09-27 00:02:00       8952.933  4323000.0 -0.001156  
2017-09-27 00:03:00       2096.655  4323000.0  0.000000  
2017-09-27 00:04:00      30398.340  4321000.0 -0.000463

In [16]:
btc.describe()

open          high           low         close         volume  \
count  2.954900e+05  2.954900e+05  2.954900e+05  2.954900e+05  295490.000000   
mean   1.164352e+07  1.165814e+07  1.162884e+07  1.164356e+07      16.387026   
std    5.109608e+06  5.120270e+06  5.099073e+06  5.109652e+06      21.895205   
min    4.318000e+06  4.318000e+06  4.318000e+06  4.318000e+06       0.000002   
25%    8.167000e+06  8.174000e+06  8.158000e+06  8.167000e+06       2.885368   
50%    9.813000e+06  9.822000e+06  9.802000e+06  9.812000e+06       9.766907   
75%    1.331100e+07  1.332600e+07  1.329900e+07  1.331100e+07      21.825489   
max    2.887900e+07  2.888500e+07  2.878400e+07  2.887900e+07     534.966763   

       volume_amount         price       log_diff  
count   2.954900e+05  2.954900e+05  295490.000000  
mean    2.398600e+08  1.164351e+07       0.000003  
std     3.782617e+08  5.109613e+06       0.002223  
min     1.014200e+01  4.318000e+06      -0.107021  
25%     2.648089e+07  8.165750e+06      -0.000579  
50%     1.041614e+08  9.812625e+06       0.000000  
75%     2.970144e+08  1.331175e+07       0.000600  
max     8.608684e+09  2.883250e+07       0.100611

### Visualization

In [27]:
daily_btc = btc.price.groupby(by=pd.Grouper(freq='d')).mean()

In [28]:
# create a new plot# create 
p = figure(
   tools="pan,box_zoom,reset,save",
   title="BTC PRICE",
   x_axis_label='date', y_axis_label='Price',
   x_axis_type='datetime', plot_width=800
)

# add some renderers
p.line(daily_btc.index, daily_btc, legend='Bitcoin', color="blue")

# show the results
show(p, notebook_handle=True)

In [34]:
daily_btc.loc[daily_btc==daily_btc.max()]

datetime_utc
2018-01-06    2.591909e+07
Freq: D, Name: price, dtype: float64

2018년 1월 6일 이전과 이후의 양상이 다르다. 나눠서 확인해보자.

1월 6일 이전

In [38]:
prev = btc.loc[:'2018-01-06']

In [39]:
prev.describe()

open          high           low         close         volume  \
count  1.349320e+05  1.349320e+05  1.349320e+05  1.349320e+05  134932.000000   
mean   1.208898e+07  1.210459e+07  1.207280e+07  1.208915e+07      15.281319   
std    6.287428e+06  6.300566e+06  6.274602e+06  6.287621e+06      23.826182   
min    4.318000e+06  4.318000e+06  4.318000e+06  4.318000e+06       0.000002   
25%    6.848000e+06  6.850000e+06  6.844000e+06  6.848000e+06       0.046804   
50%    8.968000e+06  8.974000e+06  8.958000e+06  8.966500e+06       6.167402   
75%    1.894200e+07  1.897000e+07  1.891400e+07  1.894200e+07      22.116531   
max    2.887900e+07  2.888500e+07  2.878400e+07  2.887900e+07     456.152647   

       volume_amount         price       log_diff  
count   1.349320e+05  1.349320e+05  134932.000000  
mean    2.741077e+08  1.208888e+07       0.000014  
std     4.650943e+08  6.287512e+06       0.002771  
min     1.014200e+01  4.318000e+06      -0.107021  
25%     3.210223e+05  6.847000e+06      -0.000608  
50%     6.123273e+07  8.966500e+06       0.000000  
75%     3.904028e+08  1.894150e+07       0.000658  
max     8.608684e+09  2.883250e+07       0.100611

1월 6일 이후

In [42]:
forward = btc.loc['2018-01-07':]

In [43]:
forward.describe()

open          high           low         close         volume  \
count  1.605580e+05  1.605580e+05  1.605580e+05  1.605580e+05  160558.000000   
mean   1.126915e+07  1.128294e+07  1.125573e+07  1.126910e+07      17.316256   
std    3.810536e+06  3.818452e+06  3.802596e+06  3.810356e+06      20.082736   
min    6.631000e+06  6.670000e+06  6.620000e+06  6.636000e+06       0.000138   
25%    8.829000e+06  8.840000e+06  8.817000e+06  8.828000e+06       5.755376   
50%    9.999000e+06  1.000400e+07  9.990000e+06  9.999000e+06      11.265400   
75%    1.250600e+07  1.251500e+07  1.250000e+07  1.250600e+07      21.628127   
max    2.718500e+07  2.721000e+07  2.711500e+07  2.718700e+07     534.966763   

       volume_amount         price       log_diff  
count   1.605580e+05  1.605580e+05  160558.000000  
mean    2.110784e+08  1.126923e+07      -0.000006  
std     2.823581e+08  3.810450e+06       0.001625  
min     1.000355e+03  6.646250e+06      -0.039469  
25%     5.671388e+07  8.828250e+06      -0.000560  
50%     1.173247e+08  9.997500e+06       0.000000  
75%     2.486080e+08  1.250675e+07       0.000556  
max     8.319076e+09  2.715875e+07       0.050655

**평균 상승률(Log Diff)**

2018년 1월 6일 이전 0.000014

2018년 1월 6일 이후 -0.000006

중간값은 모두 0

### 상승확률

**2018년 1월 6일 이전**

In [49]:
len(prev[prev.log_diff>0])/len(prev.log_diff)

0.48732694987104613

In [50]:
len(prev[prev.log_diff==0])/len(prev.log_diff)

0.052196662022352

In [56]:
1-0.48732694987104613-0.052196662022352

0.4604763881066018

In [126]:
# 손익비
(prev[prev.log_diff>0].mean()/abs(prev[prev.log_diff<0].mean())).log_diff

0.96263320355664961

**2018년 1월 6일 이후**

In [51]:
len(forward[forward.log_diff>0])/len(forward.log_diff)

0.4952166818221453

In [52]:
len(forward[forward.log_diff==0])/len(forward.log_diff)

0.015495957847008558

In [57]:
1-0.4952166818221453-0.015495957847008558

0.48928736033084613

In [129]:
# 손익비
(forward[forward.log_diff>0].mean()/abs(forward[forward.log_diff<0].mean())).log_diff

0.97546756620881314

승률 50%기준 손익비가 1 이상이어야 이익 (log diff 기준)

**전체기간**

In [54]:
(len(forward[forward.log_diff>0]) + len(prev[prev.log_diff>0]))/(len(forward.log_diff)+len(prev.log_diff))

0.49161392940539445

In [55]:
(len(forward[forward.log_diff==0]) + len(prev[prev.log_diff==0]))/(len(forward.log_diff)+len(prev.log_diff))

0.03225489864293208

**1분후 상승확률**

- 2018년 1월 6일 이전: 48.73%
- 2018년 1월 6일 이후: 49.52%

** 1분후 하락확률**

- 2018년 1월 6일 이전: 46.05%
- 2018년 1월 6일 이후: 48.93%

** 손익비 **

- 2018년 1월 6일 이전: 96.26%
- 2018년 1월 6일 이후: 97.55%

### 2분전 하락 & 1분전 상승시의 기초 통계

**2018년 1월 6일 이전**

In [150]:
condition = (prev.log_diff.shift(1) > 0) & (prev.log_diff.shift(2) < 0)

In [151]:
prev[condition].describe()

open          high           low         close        volume  \
count  2.640000e+04  2.640000e+04  2.640000e+04  2.640000e+04  26400.000000   
mean   1.135007e+07  1.136388e+07  1.133623e+07  1.135063e+07     12.677809   
std    6.229292e+06  6.241813e+06  6.217827e+06  6.229875e+06     21.346180   
min    4.318000e+06  4.318000e+06  4.318000e+06  4.318000e+06      0.000002   
25%    6.547000e+06  6.549750e+06  6.543000e+06  6.548000e+06      0.004315   
50%    8.411000e+06  8.414000e+06  8.402000e+06  8.410000e+06      3.178923   
75%    1.858000e+07  1.860225e+07  1.855500e+07  1.858425e+07     18.519064   
max    2.863100e+07  2.869900e+07  2.863100e+07  2.869800e+07    456.152647   

       volume_amount         price      log_diff  
count   2.640000e+04  2.640000e+04  26400.000000  
mean    2.269228e+08  1.135020e+07      0.000015  
std     4.166457e+08  6.229669e+06      0.003013  
min     1.014200e+01  4.318000e+06     -0.107021  
25%     2.329520e+04  6.546875e+06     -0.000466  
50%     2.714552e+07  8.409125e+06      0.000087  
75%     3.134252e+08  1.858031e+07      0.000818  
max     8.509205e+09  2.865925e+07      0.035063

In [152]:
len(prev[(prev.log_diff > 0)&condition])/len(prev[condition])

0.537689393939394

In [153]:
len(prev[(prev.log_diff == 0)&condition])/len(prev[condition])

0.05125

In [154]:
len(prev[(prev.log_diff < 0)&condition])/len(prev[condition])

0.41106060606060607

In [155]:
# 손익비
(prev[(prev.log_diff > 0)&condition].mean()/abs(prev[(prev.log_diff < 0)&condition].mean())).log_diff

0.77933653021613547

**2018년 1월 6일 이후**

In [156]:
condition = (forward.log_diff.shift(1) > 0) & (forward.log_diff.shift(2) < 0)

In [157]:
forward[condition].describe()

open          high           low         close        volume  \
count  2.871800e+04  2.871800e+04  2.871800e+04  2.871800e+04  28718.000000   
mean   1.120758e+07  1.122105e+07  1.119497e+07  1.120797e+07     15.915973   
std    3.761075e+06  3.768944e+06  3.754057e+06  3.761614e+06     18.329990   
min    6.682000e+06  6.700000e+06  6.661000e+06  6.670000e+06      0.000138   
25%    8.792250e+06  8.806250e+06  8.786000e+06  8.795250e+06      5.269321   
50%    9.961000e+06  9.970000e+06  9.954000e+06  9.962000e+06     10.440538   
75%    1.247600e+07  1.248900e+07  1.246500e+07  1.247600e+07     20.027904   
max    2.698600e+07  2.712500e+07  2.693200e+07  2.712000e+07    453.045367   

       volume_amount         price      log_diff  
count   2.871800e+04  2.871800e+04  28718.000000  
mean    1.940962e+08  1.120789e+07      0.000522  
std     2.642939e+08  3.761392e+06      0.001515  
min     1.000355e+03  6.678250e+06     -0.013766  
25%     5.164130e+07  8.795562e+06     -0.000129  
50%     1.077699e+08  9.961875e+06      0.000310  
75%     2.296575e+08  1.247638e+07      0.000955  
max     7.678347e+09  2.704075e+07      0.036130

In [158]:
len(forward[(forward.log_diff > 0)&condition])/len(forward[condition])

0.6755693293404833

In [159]:
len(forward[(forward.log_diff == 0)&condition])/len(forward[condition])

0.014172296120899784

In [160]:
len(forward[(forward.log_diff < 0)&condition])/len(forward[condition])

0.3102583745386169

In [161]:
# 손익비
(forward[(forward.log_diff > 0)&condition].mean()/abs(forward[(forward.log_diff < 0)&condition].mean())).log_diff

1.5853572299607936

**2분전 하락, 1분전 상승시 1분후 상승확률**

- 2018년 1월 6일 이전: 53.76% (기본: 48.73%)
- 2018년 1월 6일 이후: 67.56% (기본: 49.52%)

**2분전 하락, 1분전 상승시 1분후 하락확률**

- 2018년 1월 6일 이전: 41.11% (기본: 46.05%)
- 2018년 1월 6일 이후: 31.03% (기본: 48.93%)

**2분전 하락, 1분전 상승시 손익비**

- 2018년 1월 6일 이전: 77.93% (기본: 96.27%)
- 2018년 1월 6일 이후: 158.54% (기본: 97.54%)

### 3분전 하락, 2분 연속 상승시의 기초 통계

**2018년 1월 6일 이전**

In [175]:
condition = (prev.log_diff.shift(1) > 0) & (prev.log_diff.shift(2) > 0) & (prev.log_diff.shift(3) < 0)

In [176]:
prev[condition].describe()

open          high           low         close        volume  \
count  1.419500e+04  1.419500e+04  1.419500e+04  1.419500e+04  14195.000000   
mean   1.283668e+07  1.285316e+07  1.282034e+07  1.283797e+07     16.686055   
std    6.359590e+06  6.372120e+06  6.347932e+06  6.360746e+06     23.468183   
min    4.332000e+06  4.337000e+06  4.332000e+06  4.332000e+06      0.000004   
25%    7.294000e+06  7.300000e+06  7.278000e+06  7.292000e+06      0.464191   
50%    1.013100e+07  1.014000e+07  1.012500e+07  1.013700e+07      9.043826   
75%    1.914000e+07  1.915250e+07  1.911900e+07  1.913900e+07     24.319648   
max    2.869600e+07  2.879600e+07  2.868900e+07  2.879000e+07    365.566560   

       volume_amount         price      log_diff  
count   1.419500e+04  1.419500e+04  14195.000000  
mean    3.011624e+08  1.283704e+07     -0.000082  
std     4.594011e+08  6.360058e+06      0.002600  
min     1.828000e+01  4.333250e+06     -0.020334  
25%     3.450944e+06  7.290500e+06     -0.000524  
50%     1.088915e+08  1.013475e+07      0.000063  
75%     4.447399e+08  1.913862e+07      0.000714  
max     6.443063e+09  2.874275e+07      0.028252

In [177]:
len(prev[(prev.log_diff > 0)&condition])/len(prev[condition])

0.5318069742867206

In [178]:
len(prev[(prev.log_diff == 0)&condition])/len(prev[condition])

0.037618879887284254

In [179]:
len(prev[(prev.log_diff < 0)&condition])/len(prev[condition])

0.43057414582599507

In [180]:
# 손익비
(prev[(prev.log_diff > 0)&condition].mean()/abs(prev[(prev.log_diff < 0)&condition].mean())).log_diff

0.7203401337065537

**2018년 1월 6일 이후**

In [181]:
condition = (forward.log_diff.shift(1) > 0) & (forward.log_diff.shift(2) > 0) & (forward.log_diff.shift(3) < 0)

In [182]:
forward[condition].describe()

open          high           low         close        volume  \
count  1.940100e+04  1.940100e+04  1.940100e+04  1.940100e+04  19401.000000   
mean   1.133138e+07  1.134520e+07  1.131817e+07  1.133229e+07     16.997496   
std    3.830480e+06  3.838547e+06  3.823268e+06  3.831529e+06     19.572755   
min    6.730000e+06  6.741000e+06  6.711000e+06  6.730000e+06      0.009603   
25%    8.870000e+06  8.879000e+06  8.858000e+06  8.868000e+06      5.749920   
50%    1.003300e+07  1.004000e+07  1.002700e+07  1.003500e+07     11.140354   
75%    1.252600e+07  1.253600e+07  1.251900e+07  1.252500e+07     21.387008   
max    2.710100e+07  2.721000e+07  2.710100e+07  2.718700e+07    534.966763   

       volume_amount         price      log_diff  
count   1.940100e+04  1.940100e+04  19401.000000  
mean    2.085954e+08  1.133176e+07      0.000289  
std     2.810273e+08  3.830922e+06      0.001596  
min     9.133396e+04  6.728000e+06     -0.015813  
25%     5.667632e+07  8.867750e+06     -0.000321  
50%     1.162754e+08  1.003325e+07      0.000155  
75%     2.478807e+08  1.252725e+07      0.000740  
max     8.233700e+09  2.714975e+07      0.050655

In [183]:
len(forward[(forward.log_diff > 0)&condition])/len(forward[condition])

0.5825988351115922

In [184]:
len(forward[(forward.log_diff == 0)&condition])/len(forward[condition])

0.015411576722849337

In [185]:
len(forward[(forward.log_diff < 0)&condition])/len(forward[condition])

0.40198958816555846

In [186]:
# 손익비
(forward[(forward.log_diff > 0)&condition].mean()/abs(forward[(forward.log_diff < 0)&condition].mean())).log_diff

1.3153721150834676

**3분전 하락, 2분 연속 상승시 1분후 상승확률**

- 2018년 1월 6일 이전: 53.18% (기본: 48.73%)
- 2018년 1월 6일 이후: 58.26% (기본: 49.52%)

**3분전 하락, 2분 연속 상승시 하락확률**

- 2018년 1월 6일 이전: 43.06% (기본: 46.05%)
- 2018년 1월 6일 이후: 40.20% (기본: 48.93%)

**3분전 하락, 2분 연속 상승시 손익비**

- 2018년 1월 6일 이전: 72..03% (기본: 96.27%)
- 2018년 1월 6일 이후: 131.53% (기본: 97.54%)

#### 자동화 하자!

중복을 피하기 위해 t-1시점엔 하락, t시점부터 연속으로 상승한 경우만 테스트

#### 연속 상승

In [315]:
prev_exprs = []
for i in range(60):
    expr = '(prev.log_diff.shift({}) > 0)'.format(1)
    for j in range(i):
        if j == len(range(i))-1:
            expr += '&' + '(prev.log_diff.shift({}) < 0)'.format(j+2)
        else:
            expr += '&' + '(prev.log_diff.shift({}) > 0)'.format(j+2)
    if i == 0:
        expr += '&' + '(prev.log_diff.shift(2) < 0)'
    prev_exprs.append(expr)
prev_exprs = prev_exprs[1:]

In [316]:
forward_exprs = []
for i in range(60):
    expr = '(forward.log_diff.shift({}) > 0)'.format(1)
    for j in range(i):
        if j == len(range(i))-1:
            expr += '&' + '(forward.log_diff.shift({}) < 0)'.format(j+2)
        else:
            expr += '&' + '(forward.log_diff.shift({}) > 0)'.format(j+2)
    if i == 0:
        expr += '&' + '(prev.log_diff.shift(2) < 0)'
    forward_exprs.append(expr)
forward_exprs = forward_exprs[1:]

In [317]:
def exprs_generator(exprs,df):
    for i, expr in enumerate(exprs):
        condition = eval(expr)
        stats = df.loc[condition].log_diff.describe().to_dict()
        stats['prob'] = len(df[condition])/len(df)
        stats['expr'] = i+1
        try:
            stats['up'] = len(df[(df.log_diff > 0)&condition])/len(df[condition])
            stats['same'] = len(df[(df.log_diff == 0)&condition])/len(df[condition])
            stats['down'] = len(df[(df.log_diff < 0)&condition])/len(df[condition])
            stats['profit-to-loss'] = (df[(df.log_diff > 0)&condition].mean()/abs(df[(df.log_diff < 0)&condition].mean())).log_diff
            stats['perf'] = stats['up'] * stats['profit-to-loss']
        except ZeroDivisionError:
            stats['up'] = np.nan
            stats['same']=np.nan
            stats['down']=np.nan
            stats['profit-to-loss']=np.nan
        yield stats

In [318]:
prev_df = pd.DataFrame(exprs_generator(prev_exprs, prev))

In [319]:
prev_df.set_index('expr', inplace=True)

In [320]:
forward_df = pd.DataFrame(exprs_generator(forward_exprs, forward))

In [321]:
forward_df.set_index('expr', inplace=True)

In [322]:
prev_df = prev_df[['count', 'prob', '50%', 'mean', 'max', 'min', 'up', 'down', 'same','profit-to-loss','perf']]

In [323]:
forward_df = forward_df[['count', 'prob', '50%', 'mean', 'max', 'min', 'up', 'down', 'same','profit-to-loss','perf']]

In [324]:
prev_df = prev_df[prev_df['prob'] > 0.01]
forward_df = forward_df[forward_df['prob'] > 0.01]

In [325]:
prev_df

count      prob       50%      mean       max       min        up  \
expr                                                                        
1     26400.0  0.195654  0.000087  0.000015  0.035063 -0.107021  0.537689   
2     14195.0  0.105201  0.000063 -0.000082  0.028252 -0.020334  0.531807   
3      7549.0  0.055947  0.000164  0.000192  0.025931 -0.021716  0.590277   
4      4456.0  0.033024  0.000217  0.000375  0.031996 -0.020226  0.621185   
5      2768.0  0.020514  0.000241  0.000422  0.020978 -0.030176  0.637283   
6      1764.0  0.013073  0.000245  0.000319  0.016603 -0.016645  0.637188   

          down      same  profit-to-loss      perf  
expr                                                
1     0.411061  0.051250        0.779337  0.419041  
2     0.430574  0.037619        0.720340  0.383082  
3     0.381110  0.028613        0.867658  0.512158  
4     0.353905  0.024910        1.059382  0.658072  
5     0.342486  0.020231        1.076709  0.686168  
6     0.346939  0.015873        0.888903  0.566398

In [326]:
forward_df

count      prob       50%      mean       max       min        up  \
expr                                                                        
1     28718.0  0.178864  0.000310  0.000522  0.036130 -0.013766  0.675569   
2     19401.0  0.120835  0.000155  0.000289  0.050655 -0.015813  0.582599   
3     11303.0  0.070398  0.000203  0.000376  0.036825 -0.012548  0.612935   
4      6928.0  0.043150  0.000165  0.000288  0.018903 -0.023054  0.584729   
5      4051.0  0.025231  0.000169  0.000323  0.021767 -0.012314  0.593681   
6      2405.0  0.014979  0.000180  0.000333  0.023306 -0.015786  0.594179   

          down      same  profit-to-loss      perf  
expr                                                
1     0.310258  0.014172        1.585357  1.071019  
2     0.401990  0.015412        1.315372  0.766334  
3     0.372556  0.014509        1.400015  0.858118  
4     0.399538  0.015733        1.268761  0.741881  
5     0.390027  0.016292        1.299409  0.771434  
6     0.390852  0.014969        1.296558  0.770387

#### 연속 하락

In [298]:
prev_exprs = []
for i in range(60):
    expr = '(prev.log_diff.shift({}) < 0)'.format(1)
    for j in range(i):
        if j == len(range(i))-1:
            expr += '&' + '(prev.log_diff.shift({}) > 0)'.format(j+2)
        else:
            expr += '&' + '(prev.log_diff.shift({}) < 0)'.format(j+2)
    if i == 0:
        expr += '&' + '(prev.log_diff.shift(2) > 0)'
    prev_exprs.append(expr)
prev_exprs = prev_exprs[1:]

In [299]:
forward_exprs = []
for i in range(60):
    expr = '(forward.log_diff.shift({}) < 0)'.format(1)
    for j in range(i):
        if j == len(range(i))-1:
            expr += '&' + '(forward.log_diff.shift({}) > 0)'.format(j+2)
        else:
            expr += '&' + '(forward.log_diff.shift({}) < 0)'.format(j+2)
    if i == 0:
        expr += '&' + '(prev.log_diff.shift(2) > 0)'
    forward_exprs.append(expr)
forward_exprs = forward_exprs[1:]

In [300]:
def exprs_generator(exprs,df):
    for i, expr in enumerate(exprs):
        condition = eval(expr)
        stats = df.loc[condition].log_diff.describe().to_dict()
        stats['prob'] = len(df[condition])/len(df)
        stats['expr'] = i+1
        try:
            stats['up'] = len(df[(df.log_diff > 0)&condition])/len(df[condition])
            stats['same'] = len(df[(df.log_diff == 0)&condition])/len(df[condition])
            stats['down'] = len(df[(df.log_diff < 0)&condition])/len(df[condition])
            stats['profit-to-loss'] = (df[(df.log_diff > 0)&condition].mean()/abs(df[(df.log_diff < 0)&condition].mean())).log_diff
            stats['perf'] = stats['up'] * stats['profit-to-loss']
        except ZeroDivisionError:
            stats['up'] = np.nan
            stats['same']=np.nan
            stats['down']=np.nan
            stats['profit-to-loss']=np.nan
        yield stats

In [301]:
prev_df_down = pd.DataFrame(exprs_generator(prev_exprs, prev))

In [302]:
prev_df_down.set_index('expr', inplace=True)

In [303]:
forward_df_down = pd.DataFrame(exprs_generator(forward_exprs, forward))

In [304]:
forward_df_down.set_index('expr', inplace=True)

In [305]:
prev_df_down = prev_df_down[['count', 'prob', '50%', 'mean', 'max', 'min', 'up', 'down', 'same','profit-to-loss','perf']]

In [306]:
forward_df_down = forward_df_down[['count', 'prob', '50%', 'mean', 'max', 'min', 'up', 'down', 'same','profit-to-loss','perf']]

In [307]:
prev_df_down = prev_df_down[prev_df_down['prob'] > 0.01]
forward_df_down = forward_df_down[forward_df_down['prob'] > 0.01]

In [308]:
prev_df_down

count      prob       50%      mean       max       min        up  \
expr                                                                        
1     26271.0  0.194698 -0.000083 -0.000021  0.100611 -0.076476  0.415629   
2     14084.0  0.104379 -0.000045  0.000081  0.036830 -0.054562  0.447884   
3      7305.0  0.054138 -0.000104 -0.000097  0.064469 -0.028204  0.421355   
4      4042.0  0.029956 -0.000150 -0.000287  0.015785 -0.026903  0.405987   
5      2319.0  0.017186 -0.000137 -0.000370  0.024209 -0.034608  0.412247   

          down      same  profit-to-loss      perf  
expr                                                
1     0.536104  0.048266        1.255792  0.521944  
2     0.518674  0.033442        1.292738  0.578997  
3     0.553320  0.025325        1.139021  0.479932  
4     0.573726  0.020287        0.918686  0.372975  
5     0.570936  0.016818        0.806376  0.332426

In [309]:
forward_df_down

count      prob       50%      mean       max       min        up  \
expr                                                                        
1     28767.0  0.179169 -0.000312 -0.000538  0.016354 -0.039469  0.310251   
2     19352.0  0.120530 -0.000149 -0.000270  0.021039 -0.036741  0.404248   
3     11219.0  0.069875 -0.000180 -0.000354  0.016768 -0.021984  0.389429   
4      6706.0  0.041767 -0.000179 -0.000371  0.018008 -0.019449  0.397256   
5      3973.0  0.024745 -0.000192 -0.000396  0.022364 -0.019774  0.398943   
6      2330.0  0.014512 -0.000184 -0.000381  0.024798 -0.019090  0.407725   

          down      same  profit-to-loss      perf  
expr                                                
1     0.672715  0.017033        0.603820  0.187336  
2     0.579733  0.016019        0.788178  0.318619  
3     0.597736  0.012835        0.715636  0.278689  
4     0.592455  0.010289        0.703031  0.279283  
5     0.586459  0.014599        0.678796  0.270801  
6     0.580687  0.011588        0.728926  0.297202

In [314]:
# create a new plot
p = figure(
   tools="pan,box_zoom,reset,save",
   title="Consecutive Up & Down",
   x_axis_label='datetime', y_axis_label='Perf', plot_width=800
)

# add some renderers
p.line(prev_df.index, prev_df.perf, legend='up', color="red")
p.line(prev_df_down.index, prev_df_down.perf, legend="down", color="blue")

# show the results
show(p, notebook_handle=True)

In [327]:
# create a new plot
p = figure(
   tools="pan,box_zoom,reset,save",
   title="Consecutive Up & Down",
   x_axis_label='datetime', y_axis_label='Perf', plot_width=800
)

# add some renderers
p.line(forward_df.index, forward_df.perf, legend='up', color="red")
p.line(forward_df_down.index, forward_df_down.perf, legend="down", color="blue")

# show the results
show(p, notebook_handle=True)

활용방안
----------

3분이상 연속으로 상승한 경우  그 다음 봉도 증가할 확률이 높다.

만약 해당 코인을 보유하고 있을 경우, 그 다음 봉에 코인을 보유할지 말지에 대한 기준이 될 수 있을 것 같다.

코인을 보유하고 있지 않을 경우, 3~4분 연속으로 양봉이 나오는 시점에 매입을 고민하면 좋을 듯하다.

문제점 및 한계
----------

1. UP,DOWN만 활용했기 때문에 상대 강도를 확인하지 못한다. 10%오른것과 0.1%오른 것이 같은 비중으로 처리된다.

2. 1분뒤의 결과만 확인했기 때문에 그 이후의 사정에 대해 알려주지 못한다. 1분뒤 상승하더라도 그 이후 추세가 하락할 수 있다. 반대로 1분뒤 하락하더라도 그 이후 전반적인 추세가 역전될 수 있다. 즉 전반적인 추세를 반영하지 못하는 한계가 있다.

3. 연속 상승/연속 하락이 아니라도 그에 준하는 경우, 예를 들어 4분간 1%, 2%, -0.1%, 3%의 수익률을 기록했다고 하자. 이는 실제로는 거의 연3상승에 준하지만 실제로는 1회 상승이라고 반영된다. (노이즈)

4. 1분간의 평균 가격은 1분이 지나야 확실히 알 수 있으므로 look-forward bias가 있을 수 있다. 4분연속 상승한 것을 확인하고 매입할 때는 이미 오른 이후일 수 있다. 

추가분석
----------
1. 1분봉 뿐만이 아니라, 5분봉, 10분봉, 30분봉, 60분봉, 일봉에 대해서도 확인
2. 가격 대신, 이동평균을 이용하여 문제점 3과 같은 문제를 보완 (노이즈 제거)
3. 1분후뿐만이 아니라 그 이후의 추이에 대해서도 확인

비트코인 이외에 다른 코인들도 해봐야함.